File: guides/micronaut-security-session-database-authentication/java/src/main/java/example/micronaut/controllers/SignUpForm.java
/*
 * Copyright 2017-2023 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package example.micronaut.controllers;

import example.micronaut.constraints.PasswordMatch;
import io.micronaut.serde.annotation.Serdeable;
import io.micronaut.views.fields.annotations.InputPassword;
import jakarta.validation.constraints.NotBlank;

@PasswordMatch // <1>
@Serdeable // <2>
public record SignUpForm(@NotBlank String username, // <3>
                         @InputPassword @NotBlank String password, // <4>
                         @InputPassword @NotBlank String repeatPassword) { // <4>
}

File: guides/micronaut-security-session-database-authentication/python/src/example/micronaut/controllers/sign_up_form.py
from dataclasses import dataclass
from typing import Annotated

from jakarta.validation.constraints import NotBlank
from micronaut.core.annotation import Introspected
from micronaut.serde.annotation import Serdeable
from micronaut.views.fields.annotations import InputPassword

from ..constraints.password_match import PasswordMatch


@PasswordMatch()  # <1>
@Introspected
@Serdeable  # <2>
@dataclass
class SignUpForm:
    username: Annotated[str, NotBlank]  # <3>
    password: Annotated[str, InputPassword, NotBlank]  # <4>
    repeatPassword: Annotated[str, InputPassword, NotBlank]  # <4>
